subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskOptionsSupplier.java

```diff
@@ -0,0 +1,129 @@
+        )
+    ).collect(Collectors.toList());
+
+    /**
+     * Builds a list of implicit built-in options available for every task.
+     */
+    private static List<OptionDescriptor> getBuiltInOptions(Object target, Collection<String> reserved) {
+        List<OptionDescriptor> allBuiltInOptions = BUILT_IN_OPTIONS.stream().map(optionElement ->
+            new InstanceOptionDescriptor(target, optionElement, null, reserved.contains(optionElement.getOptionName()))
+        ).collect(Collectors.toList());
+        List<OptionDescriptor> validBuiltInOptions = new ArrayList<>();
+        for (OptionDescriptor builtInOption : allBuiltInOptions) {
+            // built-in options only enabled if they do not clash with task-declared ones
+            if (builtInOption.isClashing()) {
+                LOGGER.warn("Built-in option '{}' in task {} was disabled for clashing with another option of same name", builtInOption.getName(), target);
+            } else {
+                validBuiltInOptions.add(builtInOption);
+            }
+        }
+        return validBuiltInOptions;
+    }
+
+    /**
+     * Builds a map of generated opposite options and
+     * add pairs of mutually exclusive options to the {@link TaskOptions taskOptions}.
+     */
+    private static Map<String, OptionDescriptor> getOppositeOptions(Object target, Map<String, OptionDescriptor> options, TaskOptions taskOptions) {
+        Map<String, OptionDescriptor> oppositeOptions = new HashMap<>();
+        List<Pair<OptionDescriptor, OptionDescriptor>> mutuallyExclusiveOptions = new LinkedList<>();
+
+        for (OptionDescriptor option : options.values()) {
+            if (option instanceof InstanceOptionDescriptor) {
+                OptionElement optionElement = ((InstanceOptionDescriptor) option).getOptionElement();
+                if (optionElement instanceof BooleanOptionElement) {
+                    BooleanOptionElement oppositeOptionElement = BooleanOptionElement.oppositeOf((BooleanOptionElement) optionElement);
+                    String oppositeOptionName = oppositeOptionElement.getOptionName();
+                    if (options.containsKey(oppositeOptionName)) {
+                        LOGGER.warn("Opposite option '{}' in task {} was disabled for clashing with another option of same name", oppositeOptionName, target);
+                    } else {
+                        OptionDescriptor oppositeOption = new InstanceOptionDescriptor(target, oppositeOptionElement, null);
+                        oppositeOptions.put(oppositeOptionName, oppositeOption);
+                        mutuallyExclusiveOptions.add(Pair.of(option, oppositeOption));
+                    }
+                }
+            }
+        }
+        taskOptions.mutuallyExclusiveOptions = Collections.unmodifiableList(mutuallyExclusiveOptions);
+        return oppositeOptions;
+    }
+
+    public static TaskOptions get(Object target, OptionReader optionReader) {
```

Comment:
❕ Public method with `Object` parameter should probably have doc describing what types are appropriate for that param.